### PR TITLE
Build WMCore services and publish them to pypy, usinig to tag pushes

### LINKED ISSUE #11276: Automated packaging and upload of WMCore services to Pypi
**Impact of the new feature**
WMCore in general

**Is your feature request related to a problem? Please describe.**
As part of the migration away from RPMs, we need to start building WMCore services in a more standardized and simple way, and PyPi is a great choice for packaging and distribution of Python-based software.

**Describe the solution you'd like**
Automate the process of packaging and uploading new versions of WMCore services to PyPi. 
Expanding on the details, I think we should start with an automatic build/upload to Pypi whenever a new WMCore release is made (not a tag, but the "final" release). We need create a package for each of the services provided in WMCore, a non-exhaustive list is:
* global workqueue
* reqmgr2
* reqmon
* t0-reqmon
* wmagent
* reqmgr2ms (or each of those microservices in their own package).

This is likely to be implemented with GitHub actions, but if we decide that Jenkins would fit better our project, that is Ok as well.

**Describe alternatives you've considered**
There are many GH issues associated to this, under the "containerization" label:
https://github.com/dmwm/WMCore/issues?q=is%3Aopen+is%3Aissue+label%3Acontainerization

**Additional context**
Note that this issue is only addressing the automated packaging and distribution. Deploying our WMCore services from PyPi should be addressed in a new GH issue.

### PR DESCRIPTION
**Pull request by khurtado**
Fixes #11276

Notes:
Packages supported: `[wmagent, wmagent-devtools, wmcore, reqmon, reqmgr2, reqmgr2ms, global-workqueue, acdcserver]`

- t0-reqmon is the same as reqmon (when it comes to building the python package), so we are not building this
 Ref: 
https://github.com/cms-sw/cmsdist/blob/comp_gcc630/t0_reqmon.spec
https://github.com/cms-sw/cmsdist/blob/comp_gcc630/reqmon.spec
- t0 spec relies on 2 github repositories with packages that apparently have the same name. They are already uploading a t0 python package on pypy here: https://pypi.org/project/t0/
 https://github.com/cms-sw/cmsdist/blob/comp_gcc630/t0.spec
I'm not sure if we need to rename our package component or not, but this should be a different GH issue

#### Status
tested in a fork. 
We need to upload the pypy secret token inside the WMCore repository for this to work

#### Description
These changes create a set of github actions to automatically build and publish WMCore services (`[wmagent, wmagent-devtools, wmcore, reqmon, reqmgr2, reqmgr2ms, global-workqueue, acdcserver]`) to pypy: https://pypi.org/user/cms-oc-dmwm/

#### Is it backward compatible (if not, which system it affects?)
YES

#### External dependencies / deployment changes
https://github.com/dmwm/WMCore/pull/11324

### CODE REVIEW

**Comment by vkuznet** [comment]
For the references:
- I implemented similar GitHub Action in https://github.com/vkuznet/wmcore-builds
- I explained the whole procedure in https://github.com/dmwm/WMCore/issues/10921#issuecomment-1261201567
- And, once WMCore packages appeared in PyPi I was able to build PyPi based image, e.g. for reqmgr2 service the Dockerfile is the following:
```
FROM debian:latest
MAINTAINER Valentin Kuznetsov vkuznet@gmail.com
RUN apt-get update
RUN apt-get install -y curl libcurl4 libcurl4-openssl-dev python3-pycurl pip
RUN ln -s /usr/bin/python3 /usr/bin/python
RUN pip install reqmgr2
ENV WDIR=/data
WORKDIR $WDIR
CMD ["python3"]
```
Please note, the image is debian based, it does not have RPMs, it uses reqmgr2 PyPi package, it is 628MB image size.

But using `reqmgr2ms` package (and similar Dockefile) I get the following issue:
```
INFO: pip is looking at multiple versions of py to determine which version is compatible with other requirements. This could take a while.
Collecting py<2.0.0,>=1.4.26
  Downloading py-1.10.0-py2.py3-none-any.whl (97 kB)
  Downloading py-1.9.0-py2.py3-none-any.whl (99 kB)
  Downloading py-1.8.2-py2.py3-none-any.whl (83 kB)
  Downloading py-1.8.1-py2.py3-none-any.whl (83 kB)
  Downloading py-1.8.0-py2.py3-none-any.whl (83 kB)
  Downloading py-1.7.0-py2.py3-none-any.whl (83 kB)
  Downloading py-1.6.0-py2.py3-none-any.whl (83 kB)
  Downloading py-1.5.4-py2.py3-none-any.whl (83 kB)
  Downloading py-1.5.3-py2.py3-none-any.whl (84 kB)
  Downloading py-1.5.2-py2.py3-none-any.whl (88 kB)
  Downloading py-1.5.1-py2.py3-none-any.whl (88 kB)
  Downloading py-1.4.34-py2.py3-none-any.whl (84 kB)
  Downloading py-1.4.33-py2.py3-none-any.whl (83 kB)
  Downloading py-1.4.32-py2.py3-none-any.whl (82 kB)
  Downloading py-1.4.31-py2.py3-none-any.whl (81 kB)
  Downloading py-1.4.30-py2.py3-none-any.whl (81 kB)
  Downloading py-1.4.29-py2.py3-none-any.whl (82 kB)
  Downloading py-1.4.28-py2.py3-none-any.whl (81 kB)
  Downloading py-1.4.27-py2.py3-none-any.whl (81 kB)
  Downloading py-1.4.26.tar.gz (190 kB)
INFO: pip is looking at multiple versions of py to determine which version is compatible with other requirements. This could take a while.
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: pip is looking at multiple versions of decorator to determine which version is compatible with other requirements. This could take a while.
Collecting decorator>=3.4.2
  Downloading decorator-5.1.0-py3-none-any.whl (9.1 kB)
  Downloading decorator-5.0.9-py3-none-any.whl (8.9 kB)
  Downloading decorator-5.0.8-py3-none-any.whl (8.9 kB)
  Downloading decorator-5.0.7-py3-none-any.whl (8.8 kB)
  Downloading decorator-5.0.6-py3-none-any.whl (8.8 kB)
  Downloading decorator-5.0.5-py3-none-any.whl (8.8 kB)
  Downloading decorator-4.4.2-py2.py3-none-any.whl (9.2 kB)
  Downloading decorator-4.4.1-py2.py3-none-any.whl (9.2 kB)
  Downloading decorator-4.4.0-py2.py3-none-any.whl (8.3 kB)
  Downloading decorator-4.3.2-py2.py3-none-any.whl (9.1 kB)
  Downloading decorator-4.3.1-py2.py3-none-any.whl (8.8 kB)
  Downloading decorator-4.3.0-py2.py3-none-any.whl (9.2 kB)
  Downloading decorator-4.2.1-py2.py3-none-any.whl (9.3 kB)
  Downloading decorator-4.1.2-py2.py3-none-any.whl (9.1 kB)
  Downloading decorator-4.1.1-py2.py3-none-any.whl (9.0 kB)
  Downloading decorator-4.1.0-py2.py3-none-any.whl (9.0 kB)
  Downloading decorator-4.0.11-py2.py3-none-any.whl (8.9 kB)
  Downloading decorator-4.0.10-py2.py3-none-any.whl (9.2 kB)
  Downloading decorator-4.0.9-py2.py3-none-any.whl (22 kB)
  Downloading decorator-4.0.6-py2.py3-none-any.whl (21 kB)
  Downloading decorator-4.0.4-py2.py3-none-any.whl (8.7 kB)
  Downloading decorator-4.0.3-py2.py3-none-any.whl (8.5 kB)
  Downloading decorator-4.0.2-py2.py3-none-any.whl (8.5 kB)
  Downloading decorator-4.0.1-py2.py3-none-any.whl (8.5 kB)
  Downloading decorator-4.0.0-py2.py3-none-any.whl (8.6 kB)
  Downloading decorator-3.4.2.tar.gz (6.1 kB)
INFO: pip is looking at multiple versions of decorator to determine which version is compatible with other requirements. This could take a while.
INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. If you want to abort this run, you can press Ctrl + C to do so. To improve how pip performs, tell us what happened here: https://pip.pypa.io/surveys/backtracking
INFO: pip is looking at multiple versions of retry to determine which version is compatible with other requirements. This could take a while.
```

The bottom line is that we still have dependency chain issue with python packages which will require special attention since building and pushing packages to pypi is half way through, we need to install them to see if our builds provide valid packages.

**Comment by khurtado** [comment]
@vkuznet Wonderful! @ericvaandering is working on splitting reqmgr2ms into different pypy packages at present. Let's come back to the issue you are seeing after he manages to do #10159

**Comment by khurtado** [comment]
@germanfgv Do you have any insight on the t0 pypy package that exists already?
https://pypi.org/project/t0/

Do we need to provide support on WMCore or does T0 have that covered?

**Review by amaltaro — COMMENTED** [review]
@khurtado Kenyi, can you please update this PR with the recent changes made by Erik in this PR:
https://github.com/dmwm/WMCore/pull/11324
?

We now have separated microservice builds, and `reqmgr2ms` no longer needs to be packaged and uploaded to PyPi. @goughes can correct me here as well, in case I am missing anything.

**Comment by vkuznet** [comment]
I created a separate issue about packages which are already in pypi but for some reason which cannot be installed, see https://github.com/dmwm/WMCore/issues/11327 So far I only was able to install `reqmgr2`, but `reqmgr2ms` and `wmagent` are not installable.

**Comment by khurtado** [comment]
@amaltaro I have updated the list of pypy packages to go according to Erik changes.

File: .github/workflows/pypy_build_publish_template.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        run: |
+          cp requirements.txt requirements.wmcore.txt
+          awk "/(${{ inputs.wmcore_component }}$)|(${{ inputs.wmcore_component }},)/ {print \$1}" requirements.wmcore.txt > requirements.txt
+      - name: Build sdist
+        run: python setup.py clean sdist
+      - name: Publish component
+        uses: pypa/gh-action-pypi-publish@release/v1
+        with:
+          user: __token__
+          password: ${{ secrets.pypy_token }}
```
Comment (amaltaro) [inline .github/workflows/pypy_build_publish_template.yaml:38]: Is it supposed to be the same secrets as the yaml file above? I see they refer to different variables, is it correct?

**Review by amaltaro — CHANGES_REQUESTED** [review]
I guess GH actions are only available for this architecture:
```
    runs-on: ubuntu-latest
```
? Don't we have the possibility to use an RPM-based system (RH/Fedora/CentOS)?

Can you please squash these commits? And also ping me on slack, such that we can define the PyPi token. Thanks

File: .github/workflows/pypy_build_publish_template.yaml (hunk 1 above)
Comment (vkuznet) [reply .github/workflows/pypy_build_publish_template.yaml:38]: @amaltaro , the GitHub Actions does not run on RPM based systems, that's why we are trying to convert to pip/pypi. And, if we are doing this right it should not be matter since python code should run identically on RHAL and Debian using pypi/pip approach.

Comment (khurtado) [reply .github/workflows/pypy_build_publish_template.yaml:38]: @amaltaro I have squashed the commits. Regarding Ubuntu: We are not requiring any packages other than python 3.8 at buld-time for pypy, so I think it doesn't really make any difference.

Comment (khurtado) [reply .github/workflows/pypy_build_publish_template.yaml:38]: > Is it supposed to be the same secrets as the yaml file above? I see they refer to different variables, is it correct?

Yes. The template has its own variable name (`pypy_token`). The other file, which is the action itself, makes sure that we map `pypy_token` to the token name defined on the github settings for actions

**Review by amaltaro — APPROVED** [review]
Thanks for the clarification, Kenyi and Valentin.
From our chat on slack, I have created the Pypi token and it should be all set now. I'm creating a new tag soon to run a real test.

**Comment by amaltaro** [comment]
Haaa, nice! In 1min, we managed to build and upload everything to PyPi. Here is one example:
https://pypi.org/project/wmagent/2.1.4rc2/

I know it's unrelated to this PR, but I wonder if the PyPi release number should be `2.1.4.rc2`, or the one created `2.1.4rc2`? @goughes any thoughts? We might as well refactor our pre-releases to actually release candidates.

**Comment by khurtado** [comment]
@amaltaro I think only `.postN` and `.devN` are allowed to have a `.` after the numerical version numbers. Anything else like `rc`, `a`, `b`,  etc can't have a dot, see the PEP 440 naming convention below:

https://peps.python.org/pep-0440/#public-version-identifiers

**Comment by ericvaandering** [comment]
Not me!

Sent from a mobile device. Please excuse my brevity or transcription errors.

On Oct 5, 2022, at 2:01 PM, Kenyi Hurtado ***@***.***> wrote:

﻿

@vkuznet<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_vkuznet&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=EHaoB-POFWGrYFvPXoj1bQ&m=eJSRmgLyrKiLvQLrHU-0XTgiKqj-__yMi9qhJTFxg1ytM6TRw-dZoNNDixFnl0f9&s=tRYgCrwmetTF9GA3nNWiinWliIDw2g0qGIf3ckDrO5M&e=> Wonderful! @ericvaandering<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ericvaandering&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=EHaoB-POFWGrYFvPXoj1bQ&m=eJSRmgLyrKiLvQLrHU-0XTgiKqj-__yMi9qhJTFxg1ytM6TRw-dZoNNDixFnl0f9&s=b6D3uXR6HbnvaZoFiJvj3llUvh2Kq8BV8BNh72h-I2c&e=> is working on splitting reqmgr2ms into different pypy packages at present. Let's come back to the issue you are seeing after he manages to do #10159<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dmwm_WMCore_issues_10159&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=EHaoB-POFWGrYFvPXoj1bQ&m=eJSRmgLyrKiLvQLrHU-0XTgiKqj-__yMi9qhJTFxg1ytM6TRw-dZoNNDixFnl0f9&s=6DskT2vBNEdCQSzYMjX_Zi1BsDtfBHQhJW2siFzYHuw&e=>
